common/quotas/global/algorithm/requestweighted_test.go

```diff
@@ -0,0 +1,746 @@
+	//   with a perfect cache.
+	//   - seems not worth the complexity, but it'd be fairly easy to add if we change our minds
+	//
+	// since there is only a coarse mutex in the implementation itself, the whole aggregator forces itself
+	// to be processed serially.  if this turns out to be too costly, the easy fix is probably
+	// to shard the keys to e.g. 8 separate aggregators that can be processed concurrently.
+	//
+	// mutexes could be added to each Limit, but that could end up costing more in memory delays
+	// than the fine-grained locking gains us in concurrency flexibility.  benchmark first!
+
+	// rand and string-formatting is a non-trivial amount of load,
+	// so this is prepared up-front so it can be excluded.
+	type round struct {
+		host    Identity
+		load    map[Limit]Requests
+		keys    []Limit
+		elapsed time.Duration
+	}
+
+	rounds := make([]round, 0, b.N)
+	for i := 0; i < b.N; i++ {
+		keys := rand.Intn(keysPerHost)
+		reqs := make(map[Limit]Requests, keys)
+		for len(reqs) < keys {
+			key := Limit(fmt.Sprintf("key %d", rand.Intn(globalKeys)))
+			if _, ok := reqs[key]; ok {
+				// duplicate, just try again.
+				// dups wouldn't really invalidate the benchmark,
+				// but it's easy to avoid and should make more stable results.
+				continue
+			}
+
+			reqs[key] = Requests{
+				// values don't really matter but there's no need for them all to use the same value
+				Accepted: rand.Intn(requests),
+				Rejected: rand.Intn(requests),
+			}
+		}
+		rounds = append(rounds, round{
+			host:    Identity(fmt.Sprintf("host %d", rand.Intn(hosts))),
+			load:    reqs,
+			keys:    maps.Keys(reqs),
+			elapsed: time.Duration(rand.Int63n(updateRate.Nanoseconds())),
+		})
+	}
+
+	b.ResetTimer()
+	b.ReportAllocs()
+
+	sawnonzero := 0
+	for _, r := range rounds { // == b.N times
+		agg.Update(r.host, r.load, r.elapsed)
+		var unused map[Limit]PerSecond // ensure non-error second return for test safety
+		weights, unused := agg.HostWeights(r.host, r.keys)
+		_ = unused // ignore unused rps
+		if len(weights) > 0 {
+			// 	wrote data and later read it out, benchmark is likely functional
+			sawnonzero++
+		}
+	}
+	b.StopTimer() // not benchmarking gc, just using it to show sanity-check metrics
+
+	b.Log("N was:", b.N)
+	b.Log("gc metrics:", agg.GC()) // shows how many keys / hosts / etc we stored for manual validation
+
+	// sanity check, as "all zero" should mean something like "always looking at nonexistent keys" / bad benchmark code.
+	b.Log("nonzero results:", sawnonzero)
+	assert.True(b, b.N < 500 || sawnonzero > 0, "no non-zero result found on a large enough benchmark, likely not benchmarking anything useful")
+}
+
+func TestHashmapIterationIsRandom(t *testing.T) {
+	// is partial hashmap iteration reliably random each time so we can statistically ensure coverage,
+	// or is it relatively fixed per map?
+	// language spec is somewhat vague on the details here, so let's check.
+	//
+	// verdict: yes!  looks random each time.
```

Comment:
nice find

Reply:
I was honestly kinda surprised that it works even on the same instance.  lots of languages that have randomized maps do it with either an at-startup seed (that never changes) or a per-instance seed - Go apparently re-randomizes _every time_, which is almost certainly more costly.

I wholeheartedly approve, I like the implementation-specific-reliance-resistance this brings.  just mildly surprised they actually did it.